src/activation.rs

```diff
@@ -0,0 +1,158 @@
+use wayland_client::{
+    globals::{BindError, GlobalList},
+    protocol::{wl_seat, wl_surface},
+    Dispatch, Proxy, QueueHandle,
+};
+use wayland_protocols::xdg::activation::v1::client::{xdg_activation_token_v1, xdg_activation_v1};
+
+use crate::{
+    error::GlobalError,
+    globals::{GlobalData, ProvidesBoundGlobal},
+};
+
+/// Data attached to a token request
+pub trait RequestData: Send + Sync {
+    /// App_id of the application requesting the token, if applicable
+    fn app_id(&self) -> Option<String>;
+    /// Seat and serial of the window requesting the token, if applicable.
+    ///
+    /// *Warning*: Many compositors will issue invalid tokens for requests without
+    /// recent serials. There is no way to detect this from the client-side.
+    fn serial(&self) -> Option<(&wl_seat::WlSeat, u32)>;
+    /// Surface of the window requesting the token, if applicable.
+    ///
+    /// *Warning*: Many compositors will issue invalid tokens for requests from
+    /// unfocused surfaces. There is no way to detect this from the client-side.
+    fn surface(&self) -> Option<&wl_surface::WlSurface>;
```

Comment:
No need to use `Option` for both of them, just enforce the use of what should be passed..

Reply:
Yes there is use in requesting a token without providing these.

Reply:
I mean, what could it be, it's all non-documented. You better use `app_id` if you want to have a special logic in compositor.

Reply:
E.g. requesting a token without an active surface. Which might be valid for privileged clients.

Because it is all non-documented, we *should* actually expose everything the protocol allows instead of assuming a certain compositor policy.